driver/docker/docker.go

```diff
@@ -385,3 +390,53 @@ func generateTar(files map[string]string) (io.Reader, error) {
 
 // ConfigurationOption is an option used to customize docker driver container and host config
 type ConfigurationOption func(*container.Config, *container.HostConfig) error
+
+// inspectImage inspects the operation image and returns an object of types.ImageInspect,
+// pulling the image if not found locally
+func (d *Driver) inspectImage(ctx context.Context, image bundle.InvocationImage) (types.ImageInspect, error) {
+	ii, _, err := d.dockerCli.Client().ImageInspectWithRaw(ctx, image.Image)
+	switch {
+	case client.IsErrNotFound(err):
+		fmt.Fprintf(d.dockerCli.Err(), "Unable to find image '%s' locally\n", image.Image)
+		if err := pullImage(ctx, d.dockerCli, image.Image); err != nil {
+			return ii, err
+		}
+		if ii, _, err = d.dockerCli.Client().ImageInspectWithRaw(ctx, image.Image); err != nil {
+			return ii, errors.Wrapf(err, "cannot inspect image %s", image.Image)
+		}
+	case err != nil:
+		return ii, errors.Wrapf(err, "cannot inspect image %s", image.Image)
+	}
+
+	return ii, nil
+}
+
+// validateImageDigest validates the operation image digest, if exists, against
+// the supplied repoDigests
+func (d *Driver) validateImageDigest(image bundle.InvocationImage, repoDigests []string) error {
+	if image.Digest != "" {
+		switch count := len(repoDigests); {
+		case count == 0:
+			return fmt.Errorf("image %s has no repo digests", image.Image)
+		case count > 1:
+			return fmt.Errorf("image %s has more than one repo digest", image.Image)
+		}
+
+		repoDigest := repoDigests[0]
+		// RepoDigests are of the form 'imageName@sha256:<sha256>'; we parse out the digest itself for comparison
+		// TODO: is there a better (i.e. via docker api) way to extract the digest?
+		parts := strings.Split(repoDigest, "@")
+		if len(parts) != 2 {
+			return fmt.Errorf("unable to parse repo digest %s", repoDigest)
+		}
+		digest := parts[1]
+
+		if digest == image.Digest {
+			return nil
+		}
+		return fmt.Errorf("content digest mismatch: image %s has digest %s but the value should be %s according to the bundle file", image.Image, digest, image.Digest)
+	}
+	// TODO: if digest empty, do we want to provide a warning somehow? Spec says:
+	// If the contentDigest is not present, the runtime SHOULD report an error so the user is aware that there is no contentDigest provided.
+	return nil
```

Comment:
I think this is something that the the tool should do. In developer loops, we fully expect the digest to be left off until it has stabilized. So they (tools like porter or duffle) are in a better position to bubble up and make that warning proper.

For example, allowing it to be empty when running from a local bundle (and only displaying a warning) but perhaps displaying a stronger error if there is no digest in a published bundle.

Reply:
I agree.  I've removed this TODO and we'll leave implementation up to tools.